### PR TITLE
Binding affinity prediction

### PR DESCRIPTION
**Pull request by RMeli**
Enable binding affinity prediction alongside pose prediction with multi-output models. The total loss of multi-output models is given by
```
# Compute combined loss for pose prediction and affinity prediction
loss = pose_loss(pose_log, labels) + affinity_loss(affinities_pred, affinities)
```
where `pose_loss = nn.NLLLoss()` and `affinity_loss = PseudoHuberLoss(delta=4.0)` corresponds to a custom implementation of the pseudo-Huber loss that follows the [original Caffe implementation](https://github.com/gnina/gnina/blob/master/caffe/src/caffe/layers/affinity_loss_layer.cpp).

Unit tests are provided, but the whole code needs to be tested on real data (reproduce results in [Francoeur *et al.*](https://pubs.acs.org/doi/abs/10.1021/acs.jcim.0c00411).

This PR also exposes the following parameters (see #8):
* `--dimension`
* `--resolution`
* `--ligmolcache`
* `--recmolcache`
* `--stratify_receptor`

### CODE REVIEW

**Comment by RMeli** [comment]
Fig. 3 of Ragoza *et al.* can be reproduced by the new code as well (the pipeline is now included in `examples`, since it also show how to use `gnina-torch`).

ROC-AUC

The main difference the image in #6 is how epochs are counted, which was changed in PR #7.

**Comment by RMeli** [comment]
This PR changes yet again how epochs are counted by computing the number of batches supposed to be sampled by the given iteration scheme and manually rises a `StopIteration` exception when batches are exampled. The size of the last batch is also computed explicitly, to avoid spillage in the next epoch (which, by default, is used for padding to the default batch size).

With this new definition of an epoch, Fig. 3 of Ragoza *et al.* looks as follows:
ROC-AUC

This PR now allows to reproduce Fig. 2 of Francoeur *et al.* 

Below is the Pearson's correlation coefficient and the RMSD for the `default2018` model trained on PDBbind refine crystal structures and tested on the core set:
pearson-box
rmse-box

Below is the Pearson's correlation coefficient and the RMSD for the `default2018` model trained on PDBbind refine docked structures and tested on the core set:
pearson-box
rmse-box

Differences with the published results are affected by these factors (as well as other implementation details):
* Randomness (different random seed)
* Different number of epochs (and different early stopping)
* Different affinity loss function (pseudo-Huber loss here vs L2 loss in Francoeur *et. al.*)
* Lack of `scale` parameter in affinity loss
